lib/voxpupuli/release/rake_tasks.rb

```diff
@@ -52,3 +52,63 @@
     fail "Unable to find a CHANGELOG.md entry for the #{v} release."
   end
 end
+
+desc "Prepare a release"
+task "release:prepare" do
+  Rake::Task["release:porcelain:changelog"].invoke
+  Rake::Task["release:porcelain:reference"].invoke if File.exist?('REFERENCE.md')
+end
+
+begin
+  require 'github_changelog_generator/task'
+  require 'puppet_blacksmith'
+  GitHubChangelogGenerator::RakeTask.new "release:porcelain:changelog" do |config|
+    metadata = Blacksmith::Modulefile.new
+    config.future_release = "v#{metadata.version}" if metadata.version =~ /^\d+\.\d+.\d+$/
+    config.header = "# Changelog\n\nAll notable changes to this project will be documented in this file.\nEach new release typically also includes the latest modulesync defaults.\nThese should not affect the functionality of the module."
+    config.exclude_labels = %w{duplicate question invalid wontfix wont-fix modulesync skip-changelog}
+    config.user = metadata.metadata['name'].split(%r{[-/]}).first
+    config.user = 'voxpupuli' if config.user == "puppet"
```

Comment:
How to make the user generic?  It was hardcoded to "puppet" and we changed it in our modulesync_config, but this automatic mapping will not work in all cases, for example @opus-codium publish modules as "opuscodium" on the forge…

Reply:
we could parse it from the github url? or add a param to the rake task?

Reply:
Doing it via some parameter is probably best

Reply:
Can you elaborate on the parameter part?  The only thing I find related to "parameters" with rake would require running something like:

```
bundle exec rake 'foo[my-custom-organization]'
```

… which does not look right.

Extracting the info from the project URL makes a lot of sense since we use github_changelog_generator and can safely assume these URL point to GitHub.

Reply:
We have some parameters (https://github.com/voxpupuli/modulesync_config/blob/5544dd20712f349bded8f4ae52ffa099935d9bca/config_defaults.yml#L37-L39). The question is, how do we properly get those into the running config.

One option is parameters to Rake tasks, but like you, I find those hard to work with.

Another is to deal with it in code. [create_gch_task](https://github.com/puppetlabs/puppetlabs_spec_helper/blob/d4b2c9637917da61418c6c0b1b5aab4da0e1b25a/lib/puppetlabs_spec_helper/rake_tasks.rb#L364) is one approach. Another is a config class.

Something like
```ruby
require 'puppet_blacksmith'

class GCGConfig
  def self.user=(user)
    @user = user
  end

  def self.user
    @user || metadata['name'].split(%r{[-/]}).first
  end

  def self.metadata
    @metadata ||= Blacksmith::Modulefile.new.metadata
  end
end

puts GCGConfig.user

GCGConfig.user = 'otherwise'
puts GCGConfig.user
```

So the calling code becomes:
```ruby
require 'voxpupuli/release/rake_tasks'

GCGConfig.user = 'voxpupuli'
```

And we can continue to template this in our modulesync.

Perhaps there's also a cleaner way. https://github.com/github-changelog-generator/github-changelog-generator/blob/master/lib/github_changelog_generator/task.rb doesn't appear to have a good override mechanism.

However, now that write I this, I realize the `RakeTask` evaluates everything non-lazily. I don't know if you can use a Proc or otherwise.

Another way may be (and I don't know if this works):
```ruby
def self.user=(user)
  Rake::Task['changelog'].instance_variable_set(:user, user)
end
```

Hopefully these ideas help you.

Reply:
Nice, with some duplication of code from the GCG task generator, we can have lazy evaluation.

Reply:
As a bonus for making it lazy, I think `rake module:bump:major release:prepare` now works. Previously you needed to run it as two separate rake commands.